src/pip/_internal/cli/progress_bars.py

```diff
@@ -64,5 +92,12 @@ def get_download_progress_renderer(
     """
     if bar_type == "on":
         return functools.partial(_rich_progress_bar, bar_type=bar_type, size=size)
+    elif bar_type == "json":
+        # We don't want regular users to use this progress_bar type
+        # so only use if not a TTY
+        assert (
+            not sys.stdout.isatty()
+        ), 'The "json" progress_bar type should only be used inside subprocesses.'
+        return functools.partial(_MachineReadableProgress, size=size)
```

Comment:
Is that really a good idea? I know that @pfmoore [suggest something like this here](https://github.com/pypa/pip/pull/12084#discussion_r1228655492), but I disagree with that suggestion. 

Not showing the JSON output in TTYs makes it hard to for devs to understand what the `json` option does, because they can't really test it in their terminals. It's also the same type of "we behave differently in TTYs" that got us in this situation in the first place. I agree that spamming consoles is an issue, but I think that not allowing `json` for TTYs is the wrong approach.

I would suggest limiting how many times per second we send out progress updates. I'm not sure what it is right now, but once a second should be enough.
We could even be smart about this and use something like every 3s in TTYs to prevent spam and every 0.5s otherwise to make progress updates smoother.

---

Rate-limiting also has another advantage: performance. Right now, we log to progress to console for every downloaded chunk. I don't know how large these chunks are, but if they are relatively small (smaller than ~50KB), then logging to console X times a second could be a significant performance overhead (especially if the console output itself is slow, which may happen for a few reasons). After all, we don't want our logging to slow down the download.

Reply:
> Not showing the JSON output in TTYs makes it hard to for devs to understand what the json option does, because they can't really test it in their terminals.

Given that the only use for this option is when writing code that wraps pip and processes its output, what's so hard about starting from something that wraps pip and simply dumps the output to the screen? That will at least allow you to test you got details like IO encodings right first...

> It's also the same type of "we behave differently in TTYs" that got us in this situation in the first place.

That's just how rich works. If you disagree with it, you should be taking it up with them. As far as I'm concerned, the whole point of using libraries like rich is to take advantage of their expertise in the best way to design the features they provide. In this case, I've conceded (somewhat reluctantly) that there's a point in adding a dumb "here's the raw progress data" output, but that's very much as far as I'm willing to go. If you want to argue for a more complex solution, either get rich to add it, or find a different pip developer to support you 🙂

> I would suggest limiting how many times per second we send out progress updates.

Once again, that seems like scope creep and complexity that isn't warranted given the amount of use this feature is likely to get.

> logging to console X times a second could be a significant performance overhead

Why would it be any slower than the existing progress bar? If performance of writing a progress indicator is the issue, use `--progress-bar=off`

Reply:
>  I just don't want us to be the only ones using this feature forever.

If others find the feature useful, but limited, they can propose additions themselves.

> As I said, I don't know the chunk size.

Nor do I. That's sort of the point here, I'm reviewing and pointing out potential issues, but it's up to the PR author (or the person suggesting a change) to investigate and address the question. But having said that, I stand by my statement that if the performance of the progress reporting is the problem, switch it off. If your users complain, tell them they can have a progress report, but it will increase the download time by X% (I assume you'll test the performance impact and establish what X is, if it's that important 😉)